Add handler for local login requests for new Vue UI

jxjj opened the pull request:
This add functionality to handle local logins using the new UI via `handleLocalLoginForVueUI`.  The existing `localLogin` method is unmodified (except for detecting whether we're using the new UI).

In the VueUI handler, if the request is a `GET`, the normal `vueTemplate` will be served. `POST` requests will receive JSON responses. Session cookies are set if login is successful.

To support the vue local login, a few methods were added to our models:
- `instance`: added `isUsingVueUI` method to determine whether and instance is set to use the new UI or not
- `user_model`:  added `getUserBy` method for looking up users by username, userType, etc.
-  `Entity/User`:
	- `isExpired` which determines if a user's account is expired
	- `verifyPassword` which hashes the password (twice) and checks it against the stored value


Part of https://github.com/UMN-LATIS/elevator-ui/pull/132

On [dev.elevator.umn.edu/defaultinstance](https://dev.elevator.umn.edu/defaultinstance/) for testing

cmcfadden commented on application/models/Entity/Instance.php:
```diff
@@ -1458,4 +1458,8 @@ public function getInterfaceVersion()
         return $this->interfaceVersion;
     }
+
+    public function isUsingVueUI() {
```
Same

cmcfadden replied:
Alternatively, make this a helper method, or set a flag on the Instance_controller superclass.

jxjj replied:
moved to `instance_helper`

cmcfadden commented on application/controllers/LoginManager.php:
```diff
@@ -18,7 +18,92 @@ function force_ssl() {
+		}
+
+		// Get the data from the request body
+		$cleanJSON = $this->security->xss_clean($this->input->raw_input_stream);
```
I'd rather this be a normal post (even if it's Ajax) than a JSON bundle. We're not doing bare JSON payloads anywhere else in the codebase, so this becomes an outlier. I don't want to trust a 10 year old xss_clean method from CI to actually be robust either. Would much rather stick with traditional post fields and pull them out in the standard way - it doesn't seem like the JSON is providing any benefit here over a form?

cmcfadden commented on application/controllers/LoginManager.php:
```diff
@@ -18,7 +18,92 @@ function force_ssl() {
+
+		// if the JSON data is invalid, return an error
+		if (!$data) {
+			http_response_code(400);
```
We have a helper method for doing this. `render_json` takes a payload and optionally a status code (default is 200)

jxjj replied:
lovely!

cmcfadden commented on application/controllers/LoginManager.php:
```diff
@@ -18,7 +18,92 @@ function force_ssl() {
 	}
 
+	function handleLocalLoginForVueUI() {
+		$requestMethod = $this->input->server('REQUEST_METHOD');
```
If we're breaking this out as a separate method anyways, maybe just move the render back into the localLogin method, then make this method only for the POST result? Then you can skip this request checking.

cmcfadden commented on application/models/Entity/User.php:
```diff
@@ -663,4 +663,16 @@ public function getCsvImports()
     {
         return $this->csv_imports;
     }
```
In elevator we have sort of a two tier concept of models. There are the _model items (user_model), which are fat. Then there are the Doctrine shims, which tend to be skinny because they're mostly managed by the doctrine model generator based on the XML definitions. There are magic methods on the _models that call through to doctrine when needed. This is just a weird case because we're not actually bringing up a full user_model.

jxjj commented on application/models/User_model.php:
```diff
@@ -570,6 +570,20 @@ public function generateKeys() {
+	 * 	"userType" => "local"
+	 * ]);
+	 */
+	public function getUserBy($params) {
```
moved to `user_helper`

jxjj commented:
Thanks for the feedback. Let me know if this was the direction you're thinking.

jxjj commented on application/controllers/LoginManager.php:
```diff
@@ -10,15 +10,69 @@ public function __construct()
+		]);
+	}
+
+	function handleInvalidRequestMethod() {
```
Oh, probably don't need this with render_json

jxjj commented:
Updates:
- `localLoginAsync` is now a distinct endpoint for signing in
- removed `user_helper` file and autoload.
- inlined logic for getting user
- inlined check if user's access has expired
- moved `verifyUserPassword` as private method on the `LoginManager` controller. I chose not to inline the chunk of code since I think the intent is clearer when it's a defined as a separate method.
- moved `isUsingVueU()I` helper to a protected method `Instance_Controller`.

@cmcfadden How does this seem?

jxjj commented:
> maybe open an issue to remind us to port the other vue interface setting look ups to use that controller method?

easy peasy! just ported things over now.
